google/cloud/spanner_v1/_opentelemetry_tracing.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2016 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Manages OpenTelemetry trace creation and handling"""
+
+from contextlib import contextmanager
+
+from google.api_core.exceptions import GoogleAPICallError
+from google.cloud.spanner_v1.gapic import spanner_client
+
+try:
+    from opentelemetry import trace
+    from opentelemetry.trace.status import Status, StatusCanonicalCode
+    from opentelemetry.instrumentation.utils import http_status_to_canonical_code
+
+    HAS_OPENTELEMETRY_INSTALLED = True
+except ImportError:
+    HAS_OPENTELEMETRY_INSTALLED = False
+
+
+@contextmanager
+def trace_call(name, session, extra_attributes=None):
+    if not HAS_OPENTELEMETRY_INSTALLED:
+        # empty context manager. users will have to check if the generated value is None or a span
+        yield None
+        return
+
+    tracer = trace.get_tracer(__name__)
+
+    # base attributes that we know for every trace created
+    attributes = {
+        "db.type": "spanner",
+        "db.url": spanner_client.SpannerClient.SERVICE_ADDRESS,
```

Comment:
Users can specify different API endpoints. Currently, I'm not clear on whether we want the tracing to support using different endpoints. @hengfengli Do the other libraries use user given endpoints when tracing?

Reply:
In Go, it uses OpenCensus/trace and I don't see a way to use user given endpoints: https://github.com/googleapis/google-cloud-go/blob/master/internal/trace/trace.go.

Reply:
It seems that in Go, spans are not annotated. In the design doc, I can see that both `db.url` and `net.host.name` are endpoint urls, spanner.googleapis.com:443 by default.

Reply:
> Currently, I'm not clear on whether we want the tracing to support using different endpoints.

Not sure what you mean by this? Yes, the endpoint would be recorded as an attribute on the trace. But that doesn't affect anything about where tracing data goes, which is handled by a user-specified exporter not set up here? What is the downside of recording the endpoint here?

Reply:
> What is the downside of recording the endpoint here?

There's no downside. There are just certain cases, such as using the emulator, where the endpoint on the span wouldn't match the endpoint being used. It's a pretty minor concern though so I'm okay with leaving this as is.